### PR TITLE
get street names from osm around node

### LINKED ISSUE #860: We need a better way to suggest node names from nearby intersections/streets
Right now, it uses Mapbox layer with street names but will suggest names only if the name appears on map. We should use overpass to get nearby street names instead.

### PR DESCRIPTION
**Pull request by kaligrafy**
closes #860

### CODE REVIEW

**Comment by greenscientist** [comment]
Can you add a test with what happens if we receive a failure from overpass ?
(Overpass miss  configured or we overload the API)

File: packages/transition-backend/package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
         "format": "prettier-eslint $PWD/'src/**/*.{ts,tsx}' --write"
     },
     "dependencies": {
+        "@turf/circle": "^6.3.0",
```
Comment (tahini) [inline packages/transition-backend/package.json:30]: Is it not already in @turf/turf?

File: packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2024, Polytechnique Montreal and contributors
+ *
+ * This file is licensed under the MIT License.
+ * License text available at https://opensource.org/licenses/MIT
+ */
+import { getStreetsAroundPoint } from '../OsmGetStreetsAroundPoint';
+
+jest.mock('chaire-lib-common/lib/utils/osm/OsmOverpassDownloader', () => {
```
Comment (tahini) [inline packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts:9]: Here, you can return `downloadGeojson: jest.fn()`, then use a mock version of it in the tests. Search the code base for `jest.MockedFunction` to see many examples of how this is used. I think something like this would work.

`const mockedDownloadGeojson = OsmDownloader.downloadGeojson as jest.MockedFunction<typeof OsmDownloader.downloadGeojson>`

Then in each test, you can use mockedDownloadGeojson.mockResolvedValueOnce or mockedDownloadGeojson.mockRejectedValueOnce to test specific answers.

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 3]
```diff
@@ -75,10 +76,10 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
         serviceLocator.eventManager.on('selected.deselect.node', this.onDeselect);
         serviceLocator.eventManager.on('selected.drag.node', this.onDrag);
         serviceLocator.eventManager.on('selected.dragEnd.node', this.onDragEnd);
-        serviceLocator.eventManager.on(
+        /*serviceLocator.eventManager.on(
```
Comment (tahini) [inline packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:79]: Remove instead of comment?

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 4]
```diff
@@ -142,6 +143,47 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
             transitNodes1000mRadius: turfFeatureCollection(geojson.geometry ? [geojson] : []),
             transitNodesRoutingRadius: turfFeatureCollection(geojson.geometry ? [geojson] : [])
         });
+        serviceLocator.socketEventManager.emit(
```
Comment (tahini) [inline packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:146]: Is this the right place to put this code?

**Review by tahini — CHANGES_REQUESTED** [review]

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 3 above)
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:79]: removed

File: packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts (hunk 2 above)
Comment (kaligrafy) [reply packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts:11]: fixed

File: packages/transition-backend/package.json (hunk 1 above)
Comment (kaligrafy) [reply packages/transition-backend/package.json]: fixed

File: packages/transition-backend/src/api/osm.socketRoutes.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2024, Polytechnique Montreal and contributors
+ *
+ * This file is licensed under the MIT License.
+ * License text available at https://opensource.org/licenses/MIT
+ */
+import geobuf from 'geobuf';
+import Pbf from 'pbf';
```
Comment (tahini) [inline packages/transition-backend/src/api/osm.socketRoutes.ts:8]: Need this import?

File: packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import * as Status from 'chaire-lib-common/lib/utils/Status';
+
+jest.mock('chaire-lib-common/lib/utils/osm/OsmOverpassDownloader', () => {
+    return { downloadGeojson: jest.fn() };
+});
+
+describe('getStreetsAroundPoint', () => {
+    const mockedDownloadGeojson = OsmDownloader.downloadGeojson as jest.MockedFunction<typeof OsmDownloader.downloadGeojson>;
+    it('should return mocked streets around a given point', async () => {
+        mockedDownloadGeojson.mockResolvedValueOnce({
```
Comment (tahini) [inline packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts:18]: Maybe a test with a valid, but unexpected geojson (like no street name)

File: packages/transition-backend/src/api/osm.socketRoutes.ts (hunk 5 above)
Comment (kaligrafy) [reply packages/transition-backend/src/api/osm.socketRoutes.ts:8]: removed

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 3 above)
Comment (tahini) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:79]: not removed?

File: packages/chaire-lib-common/src/config/osm/overpassQueries/streets.overpassql [hunk 7]
```diff
@@ -0,0 +1,8 @@
+
+[out:json][timeout:25];
```
Comment (greenscientist) [inline packages/chaire-lib-common/src/config/osm/overpassQueries/streets.overpassql:2]: What's the rationale for a 25 secondes timeout ?

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 8]
```diff
@@ -142,6 +143,43 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
...
+                                .filter((f) => f.properties?.name !== undefined)
+                                .map((f) => f.properties?.name)
+                        );
+                        const possibleIntersectionNames: string[] = [];
+                        for (let i = 0, countI = uniqStreetNames.length; i < countI; i++) {
+                            for (let j = i + 1, countJ = uniqStreetNames.length; j < countJ; j++) {
+                                const streetName1 = uniqStreetNames[i];
+                                const streetName2 = uniqStreetNames[j];
+                                if (streetName1 !== streetName2) {
+                                    possibleIntersectionNames.push(`${streetName1} / ${streetName2}`);
```
Comment (tahini) [inline packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:166]: Should you push the 2 possible permutations? like Villeray/St-Michel, St-Michel/Villeray?

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 9]
```diff
@@ -142,6 +143,43 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
...
             transitNodesRoutingRadius: turfFeatureCollection(geojson.geometry ? [geojson] : [])
         });
+        serviceLocator.socketEventManager.emit(
+            'osm.streetsAroundPoint',
+            this.state.object.toGeojson(),
+            100,
+            (status: Status.Status<GeoJSON.Feature[]>) => {
+                if (Status.isStatusOk(status)) {
+                    const streetsGeojsonFeatures: GeoJSON.Feature[] = status.result;
+                    if (streetsGeojsonFeatures.length < 10) {
```
Comment (greenscientist) [inline packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:153]: probably need an else (and also a specific test for this condition) with a log to know what is the error if we get it

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 10]
```diff
@@ -142,6 +143,43 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
...
+                            for (let j = i + 1, countJ = uniqStreetNames.length; j < countJ; j++) {
+                                const streetName1 = uniqStreetNames[i];
+                                const streetName2 = uniqStreetNames[j];
+                                if (streetName1 !== streetName2) {
+                                    possibleIntersectionNames.push(`${streetName1} / ${streetName2}`);
+                                }
+                            }
+                        }
+                        // add the single street name if only one street name was found:
+                        if (possibleIntersectionNames.length === 0 && uniqStreetNames.length === 1) {
```
Comment (greenscientist) [inline packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:171]: Would it be worthwhile to return an address here ? STM does that for places without intersection in their stop name. 
Could be just a TODO for now. Or an option to the function.

File: packages/transition-backend/src/services/osm/OsmGetStreetsAroundPoint.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import { DataGeojson } from 'chaire-lib-common/lib/tasks/dataImport/data/dataGeojson';
+import osmDownloader from 'chaire-lib-common/lib/utils/osm/OsmOverpassDownloader';
+import streetsXMLQuery from 'chaire-lib-common/lib/config/osm/overpassQueries/streets';
+import * as Status from 'chaire-lib-common/lib/utils/Status';
+
+export const getStreetsAroundPoint = async function (
+    aroundPoint: GeoJSON.Feature<GeoJSON.Point>,
+    radiusMeters: number
+): Promise<Status.Status<GeoJSON.Feature[]>> {
+    const circleGeojson = turfCircle(aroundPoint, radiusMeters, { units: 'meters' });
```
Comment (greenscientist) [inline packages/transition-backend/src/services/osm/OsmGetStreetsAroundPoint.ts:17]: What happen if radius is negative. 
(We probably need a test for that also)

**Review by greenscientist — COMMENTED** [review]
Also, I was not sure if it was there, but do we have tests for the case with actual intersection, with ones without and ones with only one street?

**Comment by greenscientist** [comment]
Sorry to pile on with comments, but it's a simple small new feature, it's a good opportunity to have proper architectures and tests!

File: packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts (hunk 6 above)
Comment (kaligrafy) [reply packages/transition-backend/src/services/osm/__tests__/OsmGetStreetsAroundPoint.test.ts:18]: this service does not check for street name, only for streets around node. The check for name is in the  NodeEditForm, which is not yet tested...

File: packages/chaire-lib-common/src/config/osm/overpassQueries/streets.overpassql (hunk 7 above)
Comment (kaligrafy) [reply packages/chaire-lib-common/src/config/osm/overpassQueries/streets.overpassql:2]: I don't know, what would you put there?

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 8 above)
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:166]: done

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 9 above)
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx]: This part is changed, we now check for intersections with turf and order them by distance

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 10 above)
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx]: Addresses are rare in osm, and we would need to use Photon maybe?

Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx]: added a todo

File: packages/transition-backend/src/services/osm/OsmGetStreetsAroundPoint.ts (hunk 11 above)
Comment (kaligrafy) [reply packages/transition-backend/src/services/osm/OsmGetStreetsAroundPoint.ts:20]: Added a validation for radius <= 0

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 3 above)
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx]: pushed now

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 4 above)
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx]: NO, I moved it to componentDidMount and onValueChange

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 12]
```diff
@@ -208,6 +207,126 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
         });
     };
 
+    private fillAutocompleteName = async () => {
```
Comment (tahini) [inline packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:210]: Just saying, if you put it in a nice utility function somewhere in a service instead of the component (and this is not really related to a react component's structure), then you can add as many unit tests as you can think of!! The autocomplete list for the code is well unit tested in its service file...

And if you put it in a utility function, it WILL be used again elsewhere at some point...

**Review by tahini — CHANGES_REQUESTED** [review]

File: packages/chaire-lib-common/src/config/osm/overpassQueries/streets.overpassql (hunk 7 above)
Comment (kaligrafy) [reply packages/chaire-lib-common/src/config/osm/overpassQueries/streets.overpassql]: added a comment and changed to 10 seconds

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx (hunk 12 above)
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:207]: done!

File: packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
+import { v4 as uuidV4 } from 'uuid';
```
Comment (tahini) [inline packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts:1]: Add copyright header here

File: packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        const node = new Node({
+            id: uuidV4(),
+            code: '123',
+            geography: { type: 'Point' as const, coordinates: [-73.9975, 45.0005] },
+        }, true);
+        const maxRadiusMeters = 100;
+
+        const result = await proposeNames(mockSocket, node, maxRadiusMeters);
+
+        const distance1 = turfDistance(turfPoint([-73.9975, 45.0005]), turfPoint([-73.9970, 45.0010]));
```
Comment (tahini) [inline packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts:54]: Ideally, instead of hardcoding the values here, you could have put the coordinates in variables in the test and use them here and in the features above. Now if you change one, you'll need to update all the tests and figure out why they fail, if they fail, if they don't just silently do nothing...

**Review by tahini — CHANGES_REQUESTED** [review]

File: packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts (hunk 13 above)
Comment (kaligrafy) [reply packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts:1]: added

File: packages/transition-backend/src/services/osm/OsmGetStreetsAroundPoint.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+ * This file is licensed under the MIT License.
+ * License text available at https://opensource.org/licenses/MIT
+ */
+import { circle as turfCircle } from '@turf/turf';
+import { DataGeojson } from 'chaire-lib-common/lib/tasks/dataImport/data/dataGeojson';
+import osmDownloader from 'chaire-lib-common/lib/utils/osm/OsmOverpassDownloader';
+import streetsXMLQuery from 'chaire-lib-common/lib/config/osm/overpassQueries/streets';
+import * as Status from 'chaire-lib-common/lib/utils/Status';
+
+export const getStreetsAroundPoint = async function (
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-backend/src/services/osm/OsmGetStreetsAroundPoint.ts:13]: added

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 16]
```diff
@@ -208,6 +204,39 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
...
+        /* TODO: enable progress event. Right now in Chrome, the result appears before the progress event completion is emitted.
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:208]: Same problem, tested...

File: packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx [hunk 17]
```diff
@@ -70,15 +72,15 @@ class TransitNodeEdit extends SaveableObjectForm<Node, NodeFormProps, NodeFormSt
     }
 
     componentDidMount() {
+        console.log('mounting');
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-frontend/src/components/forms/node/TransitNodeEdit.tsx:75]: removed

File: packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts (hunk 14 above)
Comment (kaligrafy) [reply packages/transition-common/src/services/nodes/__tests__/NodeGeographyUtils.test.ts:69]: coordinates of points are different, but I added variables for streets

**Review by tahini — APPROVED** [review]
I don't know what's happening with docker. Do @greenscientist  trick of adding an extra space in the commit message and try again?

**Comment by kaligrafy** [comment]
> I don't know what's happening with docker. Do @greenscientist trick of adding an extra space in the commit message and try again?

Didn't work...
